### PR TITLE
Bugfix/s3 website configuration

### PR DESCRIPTION
**Pull request by jack1902**
### Description of your changes

Slack Thread: https://crossplane.slack.com/archives/CEF5N8X08/p1620143654284900

Fixes https://github.com/crossplane/provider-aws/issues/650

I have:

- [x] Read and followed Crossplane's [contribution process]
- [x] Run `make reviewable test` to ensure this PR is ready for review.

To Note, i have removed `protocol` as a required parameter because even the description states it defaults to what the inbound request is yet the Spec states its required which goes against the given description

### How has this code been tested

[contribution process]: https://git.io/fj2m9

- I have been trying to use the website configuration on the bucket resource within a composition. I noticed that it worked if i passed routing rules but didn't when i only passed the index and error documents (things i have already done before in terraform but i am moving the bucket creation to crossplane 🚀 )

- `make test` - ✅ 
- `make e2e` - ✅ 
Screenshot 2021-05-04 at 21 09 59

- `make reviewable` - ✅ 
- `make` - ✅

### CODE REVIEW

File: package/crds/s3.aws.crossplane.io_buckets.yaml [hunk 1]
```diff
@@ -885,7 +885,6 @@ spec:
                             type: string
                         required:
                         - hostName
-                        - protocol
```
Comment (muvaf) [inline package/crds/s3.aws.crossplane.io_buckets.yaml:888]: CRDs are automatically generated. You'll need to make [this field](https://github.com/crossplane/provider-aws/blob/master/apis/s3/v1beta1/websiteConfiguration_types.go#L64) 
```
Protocol *string `json:"protocol,omitempty"`
```
 to make it non-required then fix where it's used.

**Review by muvaf — COMMENTED** [review]
Hi @jack1902 ! Congrats on your first contribution! One small caveat and I think we're good to go!

File: package/crds/s3.aws.crossplane.io_buckets.yaml (hunk 1 above)
Comment (jack1902) [reply package/crds/s3.aws.crossplane.io_buckets.yaml:888]: How do i regen these files? Since i have made a change to the overall structs

Comment (muvaf) [reply package/crds/s3.aws.crossplane.io_buckets.yaml:888]: you can run `make reviewable`

Comment (jack1902) [reply package/crds/s3.aws.crossplane.io_buckets.yaml:888]: Running now, will push once its done (i forgot about that command as i ran `test` and `e2e` only 🤦 )

Comment (jack1902) [reply package/crds/s3.aws.crossplane.io_buckets.yaml:888]: Ran it locally, no difference in output (assuming this means my hand-crafted changes are correct)

File: pkg/controller/s3/bucket/websiteConfig.go [hunk 2]
```diff
@@ -139,28 +139,37 @@ func GenerateWebsiteConfiguration(config *v1beta1.WebsiteConfiguration) *awss3.W
...
 		}
 	}
-	wi.RoutingRules = make([]awss3.RoutingRule, len(config.RoutingRules))
-	for i, rule := range config.RoutingRules {
-		rr := awss3.RoutingRule{
-			Redirect: &awss3.Redirect{
+	if config.RoutingRules != nil {
+		wi.RoutingRules = make([]awss3.RoutingRule, len(config.RoutingRules))
+		for i, rule := range config.RoutingRules {
+			redirect := &awss3.Redirect{
```
Comment (muvaf) [inline pkg/controller/s3/bucket/websiteConfig.go:150]: Why did you need to switch from initializing with `RoutingRule` to `Redirect`? The former seems more straightforward to read to me.

File: pkg/controller/s3/bucket/websiteConfig.go [hunk 3]
```diff
@@ -139,28 +139,37 @@ func GenerateWebsiteConfiguration(config *v1beta1.WebsiteConfiguration) *awss3.W
...
+		}
+		if config.RedirectAllRequestsTo.Protocol != "" {
+			wi.RedirectAllRequestsTo.Protocol = awss3.Protocol(config.RedirectAllRequestsTo.Protocol)
 		}
 	}
-	wi.RoutingRules = make([]awss3.RoutingRule, len(config.RoutingRules))
-	for i, rule := range config.RoutingRules {
-		rr := awss3.RoutingRule{
-			Redirect: &awss3.Redirect{
+	if config.RoutingRules != nil {
```
Comment (muvaf) [inline pkg/controller/s3/bucket/websiteConfig.go:147]:
```suggestion
	if len(config.RoutingRules) != 0 {
```

This is usually a safer check for slices.

**Review by muvaf — COMMENTED** [review]
Just small comments. Make sure to squash the commits when you're done so that I can merge right away :)

File: pkg/controller/s3/bucket/websiteConfig.go (hunk 2 above)
Comment (jack1902) [reply pkg/controller/s3/bucket/websiteConfig.go:150]: i did this becaue i wasn't 100% sure if i could do `rr.Redirect.Protocol` but i realise now that i can do that (i'm a bit rusty in Go as i haven't written it in ages)

File: pkg/controller/s3/bucket/websiteConfig.go (hunk 3 above)
Comment (jack1902) [reply pkg/controller/s3/bucket/websiteConfig.go]: done

File: pkg/controller/s3/bucket/websiteConfig.go (hunk 2 above)
Comment (jack1902) [reply pkg/controller/s3/bucket/websiteConfig.go]: updated

**Review by muvaf — APPROVED** [review]
Thanks @jack1902 !
